IQSS/dataverse
# remove invalid characters from input #8018

**What this PR does / why we need it**:

Currently users can input invalid characters into the database via both the GUI and the API. This pull request strips them out before they are written to the database.

**Which issue(s) this PR closes**:

Closes #8018

**Special notes for your reviewer**:

- Should we be handling more special characters? If so, which ones? ~~If you enter all of https://github.com/minimaxir/big-list-of-naughty-strings/blob/master/blns.txt as a description we have the same problem of not being able to edit the dataset. The error looks something like this in Chrome: `error on line 13 at column 71: Input is not proper UTF-8, indicate encoding ! Bytes: 0xEF 0xBF 0xBE 0x26`~~ (fixed as of 2d8fe44)
- Any ideas for enforcing this at the database level as well? The pull request uses Bean Validation.

**Suggestions on how to test this**:

- Try entering special characters like (form feed or start of text) into dataset fields such as description or title. See the SQL script or DatasetFieldValidator.java for the full list of characters that are being removed.
- Try creating or editing a dataset via API with the same special characters.

The special characters should be removed before hitting the database as in the screenshots below.

![Screen Shot 2021-11-15 at 3 38 01 PM](https://user-images.githubusercontent.com/21006/141851314-3099c835-86bc-4dc7-8e6b-5a360ea09577.png)
![Screen Shot 2021-11-15 at 3 38 31 PM](https://user-images.githubusercontent.com/21006/141851313-3f0d4522-e454-4d36-ae19-c09b83a584fa.png)

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

No.

**Is there a release notes update needed for this change?**:

Yes, included.

**Additional documentation**:

I didn't include any but perhaps we could mention under "Adding a New Dataset" that invalid characters will be removed: https://guides.dataverse.org/en/5.8/user/dataset-management.html#adding-a-new-dataset
